src/Storages/MergeTree/KeyCondition.cpp

```diff
@@ -69,6 +70,8 @@ String extractFixedPrefixFromLikePattern(const String & like_pattern)
             case '%':
                 [[fallthrough]];
             case '_':
+                if (perfect_prefix_match && std::find_if(pos+1, end, [](const char c) { return c != '%' && c != '_'; }) != end)
```

Comment:
Unrelated: The `break` in l. 84 isn't needed.

Reply:
This should be rewritten for clarity and symmetry:

```cpp
if (perfect_prefix)
{
    bool hasLiteralsAfterFirstWildcard = (std::find_if(pos+1, end, [](auto c) { return !(c == '%' || c == '_'); }) != end); }
    return hasLiteralsAfterFirstWildcard ? "" : fixed_prefix;
}
else
    return fixed_prefix;
```

Reply:
Question: What happens to the correctness of NOT LIKE using the primary key if there exist masked wildcards after the first wildcard, e.g. `NOT LIKE 'abc%\_'`?

Reply:
Hi, thanks for this question, I found that NOT LIKE can only be used in key condition if the prefix is `prefix%`. For example, if the predicate is `s NOT LIKE 'a_'`, the current logic will construct a key condition `s not in ['a', 'b')`, which is wrong because `s = 'abb'` is in `['a', 'b')` but still satisfy `s NOT LIKE 'a_'`.